File: solutions/cpp/875/solution.cpp
int minEatingSpeed(vector<int> &piles, int h) {
    int left = 1, right = 0;
    for (int bananas : piles) {
        right = max(right, bananas)
    }

    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (check(piles, mid, h)) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return left;
}

bool check(vector<int> &piles, int k, int limit) {
    long h = 0;
    for (double bananas : piles) {
        h += ceil(bananas / k);
    }
    return h <= limit;
}

File: solutions/golang/875/solution.go
var limit int
func minEatingSpeed(piles []int, h int) int {
    limit = h
    left := 1
    right := 0
    for _, bananas := range piles {
        right = max(right, bananas)
    }

    for left <= right {
        mid := left + (right - left) / 2
        if check(mid, piles) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return left
}

func check(k int, piles []int) bool {
    hours := 0
    fk := float64(k)
    for _, bananas := range piles {
        fbananas := float64(bananas)
        hours += int(math.Ceil(fbananas / fk))
    }

    return hours <= limit
}
